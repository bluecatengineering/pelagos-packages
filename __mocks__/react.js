module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
