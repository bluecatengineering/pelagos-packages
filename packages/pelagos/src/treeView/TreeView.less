@import '../../less/spacing';
@import '../../less/utils';

@layer pelagos {
	.TreeView {
		margin: 0;
		padding: 0;
		list-style: none;

		&__node {
			outline: none;
			cursor: pointer;
		}

		&__label {
			flex-direction: row;
			align-items: center;
			height: 32px;
			padding-right: @sp-16;
			color: var(--text-secondary);
			transition:
				color 0.15s ease-out,
				background-color 0.15s ease-out;

			.TreeView--xs & {
				height: 24px;
			}

			&:hover {
				color: var(--text-primary);
				background-color: var(--layer-hover);
			}

			:focus-visible > & {
				@focus-ring();
			}

			[aria-selected='true'] > & {
				position: relative;
				color: var(--text-primary);
				background-color: var(--layer-selected);

				&:hover {
					background-color: var(--layer-selected-hover);
				}

				&::before {
					content: '';
					position: absolute;
					left: 0;
					top: 0;
					bottom: 0;
					width: 4px;
					background-color: var(--border-interactive);
				}
			}
		}

		&__iconWrapper {
			display: flex;
			align-items: center;
			justify-content: center;
			width: 16px;
			height: 16px;
			margin-right: @sp-08;
		}

		&__icon {
			color: var(--icon-secondary);
			transition:
				color 0.15s ease-out,
				transform 0.15s ease-out;

			[aria-selected='true'] &,
			.TreeView__label:hover & {
				color: var(--icon-primary);
			}
		}

		&__loading {
			font-size: 2.5px;
		}

		&--expanded {
			transform: rotate(90deg);
		}

		&__text {
			@base-ellipsis();
		}

		&__children {
			margin: 0;
			padding: 0;
			list-style: none;
		}
	}
}
