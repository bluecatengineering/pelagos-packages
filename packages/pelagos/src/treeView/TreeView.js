import {Children, useEffect, useRef, useState} from 'react';
import PropTypes from 'prop-types';

import TreeViewContext from './TreeViewContext';
import TreeNodeContext from './TreeNodeContext';

import './TreeView.less';

const nodeContextValue = {level: 0, padding: 0, parentPath: []};

const getFocused = (selected, children) => {
	if (selected?.length) {
		return selected;
	}
	if (children) {
		return [Children.toArray(children)[0].props.id];
	}
	return [];
};

/** A component implementing the tree view pattern. */
const TreeView = ({className, size, selected, children, onSelect, ...props}) => {
	const rootRef = useRef(null);
	const [focused, setFocused] = useState(() => getFocused(selected, children));
	useEffect(() => {
		if (!rootRef.current.contains(document.activeElement)) {
			setFocused(getFocused(selected, children));
		}
	}, [children, selected]);
	return (
		<TreeViewContext.Provider value={{selected, focused, setFocused, onSelect}}>
			<TreeNodeContext.Provider value={nodeContextValue}>
				<ul
					{...props}
					className={`TreeView TreeView--${size}${className ? ` ${className}` : ''}`}
					role="tree"
					ref={rootRef}>
					{children}
				</ul>
			</TreeNodeContext.Provider>
		</TreeViewContext.Provider>
	);
};

TreeView.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The node size. */
	size: PropTypes.oneOf(['sm', 'xs']),
	/** The ID of the selected node. */
	selected: PropTypes.arrayOf(PropTypes.string),
	/** The child nodes. */
	children: PropTypes.node,
	/** Function invoked when a node is selected. */
	onSelect: PropTypes.func,
};

TreeView.defaultProps = {
	size: 'sm',
};

export default TreeView;
