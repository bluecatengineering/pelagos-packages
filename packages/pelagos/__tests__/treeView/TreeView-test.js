import {useEffect, useRef, useState} from 'react';
import {shallow} from 'enzyme';

import TreeView from '../../src/treeView/TreeView';

jest.unmock('../../src/treeView/TreeView');

const anyFunction = expect.any(Function);

global.document = {activeElement: 'active-element'};

describe('TreeView', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<TreeView>children</TreeView>);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when optional properties are set', () => {
			const wrapper = shallow(
				<TreeView className="TestClass" size="xs" selected="test-id">
					children
				</TreeView>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('initializes focused to selected when specified', () => {
			shallow(<TreeView selected="test-id" />);
			expect(useState.mock.calls).toEqual([[anyFunction]]);
			expect(useState.mock.calls[0][0]()).toBe('test-id');
		});

		it('initializes focused to the id of the first child when selected is not specified', () => {
			shallow(
				<TreeView>
					<div id="child-id" />
				</TreeView>
			);
			expect(useState.mock.calls).toEqual([[anyFunction]]);
			expect(useState.mock.calls[0][0]()).toBe('child-id');
		});

		it('adds an effect which calls setFocused when selected changes', () => {
			const children = <div id="child-id" />;
			const contains = jest.fn();
			const setFocused = jest.fn();
			useRef.mockReturnValueOnce({current: {contains}});
			useState.mockReturnValueOnce(['test-id', setFocused]);
			shallow(<TreeView selected="other-id">{children}</TreeView>);
			expect(useEffect.mock.calls[0]).toEqual([anyFunction, [children, 'other-id']]);

			useEffect.mock.calls[0][0]();
			expect(contains.mock.calls).toEqual([['active-element']]);
			expect(setFocused.mock.calls).toEqual([['other-id']]);
		});

		it('adds an effect which calls setFocused when selected changes to null', () => {
			const children = <div id="child-id" />;
			const contains = jest.fn();
			const setFocused = jest.fn();
			useRef.mockReturnValueOnce({current: {contains}});
			useState.mockReturnValueOnce(['test-id', setFocused]);
			shallow(<TreeView selected={null}>{children}</TreeView>);
			expect(useEffect.mock.calls[0]).toEqual([anyFunction, [children, null]]);

			useEffect.mock.calls[0][0]();
			expect(contains.mock.calls).toEqual([['active-element']]);
			expect(setFocused.mock.calls).toEqual([['child-id']]);
		});

		it('adds an effect which does not call setFocused when a node in the tree has focus', () => {
			const children = <div id="child-id" />;
			const contains = jest.fn().mockReturnValue(true);
			const setFocused = jest.fn();
			useRef.mockReturnValueOnce({current: {contains}});
			useState.mockReturnValueOnce(['test-id', setFocused]);
			shallow(<TreeView selected="other-id">{children}</TreeView>);
			expect(useEffect.mock.calls[0]).toEqual([anyFunction, [children, 'other-id']]);

			useEffect.mock.calls[0][0]();
			expect(contains.mock.calls).toEqual([['active-element']]);
			expect(setFocused.mock.calls).toEqual([]);
		});
	});
});
