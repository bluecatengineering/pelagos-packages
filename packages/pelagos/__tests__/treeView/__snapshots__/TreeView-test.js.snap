// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView rendering renders expected elements 1`] = `
<Context.Provider
  value={
    {
      "focused": [Function],
      "onSelect": undefined,
      "selected": undefined,
      "setFocused": [MockFunction],
    }
  }
>
  <Context.Provider
    value={
      {
        "level": 0,
        "padding": 0,
      }
    }
  >
    <ul
      className="TreeView TreeView--sm"
      role="tree"
    >
      children
    </ul>
  </Context.Provider>
</Context.Provider>
`;

exports[`TreeView rendering renders expected elements when optional properties are set 1`] = `
<Context.Provider
  value={
    {
      "focused": [Function],
      "onSelect": undefined,
      "selected": "test-id",
      "setFocused": [MockFunction],
    }
  }
>
  <Context.Provider
    value={
      {
        "level": 0,
        "padding": 0,
      }
    }
  >
    <ul
      className="TreeView TreeView--xs TestClass"
      role="tree"
    >
      children
    </ul>
  </Context.Provider>
</Context.Provider>
`;
